src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs

```diff
@@ -92,7 +99,13 @@ private async Task AnalyzeDocumentForKindAsync(TextDocument document, AnalysisKi
                     foreach (var (analyzer, state) in nonCachedAnalyzersAndStates)
                     {
                         var computedData = await ComputeDocumentAnalysisDataAsync(executor, analyzer, state, logTelemetry, cancellationToken).ConfigureAwait(false);
-                        PersistAndRaiseDiagnosticsIfNeeded(computedData, analyzer, state);
+
+                        PersistAndAddDiagnosticsArgsIfNeeded(ref argsBuilder.AsRef(), computedData, analyzer, state);
+
+                        if (argsBuilder.Count > 0)
+                        {
+                            AnalyzerService.RaiseDiagnosticsUpdated(argsBuilder.ToImmutableAndClear());
```

Comment:
>AnalyzerService.RaiseDiagnosticsUpdated(argsBuilder.ToImmutableAndClear()); [](http://example.com/codeflow?start=28&length=75)

Wanting to make sure this needs to be done inside the loop

Reply:
➡️ Yes, we don't want to delay the presentation of diagnostics from analyzer A until some other analyzer B completes.

Reply:
Makes sense, thanks for the explanation

Reply:
Probably a dumb question, but would it make sense to try to utilize AsyncBatchingWorkQueue to allow these to get batched together if they do completely quickly? Another dumb question is why these are awaited serially, and not kicked off in a parallel fashion? @mavasani?

Reply:
> Probably a dumb question, but would it make sense to try to utilize AsyncBatchingWorkQueue to allow these to get batched together if they do completely quickly?

Potentially, yes. In DefaultDiagnosticAnalyzerService.cs I added a comment to that effect.

> Another dumb question is why these are awaited serially, and not kicked off in a parallel fashion?

I'm not sure, but it's outside the scope of this change.

Reply:
> Another dumb question is why these are awaited serially, and not kicked off in a parallel fashion?

There is some history here. Before we moved core analyzer execution pieces to OOP, we wanted to make sure that we don't stress the devenv process by running analyzers concurrently as this was part of continuous background analysis. After moving the analyzers to OOP, we do execute all the analyzers in OOP process in parallel. Just the requests to fetch the analysis result still happens serially here, which can potentially be changed but will not likely lead to any significant performance benefit. Note that we do special case the compiler analyzer though and execute it in isolation upfront, rest of the analyzers execute concurrently. For example, see below: https://github.com/dotnet/roslyn/blob/49e5f55e2d4e71618109cda319f2eedc39c2dcec/src/Features/LanguageServer/Protocol/Features/Diagnostics/DocumentAnalysisExecutor.cs#L216-L249